tools/run-fbp-tests

```diff
@@ -67,146 +162,162 @@ def collect_tests(tests):
+        "passed": 0,
+        "failed": 0,
+        "skipped": 0
+    }
+
+    skips = get_skip_list(args)
+    if args.valgrind:
+        skips.append(FbpTest.SKIP_VALGRIND)
 
-        out, code = sh(cmd_prefix + [runner, file], cwd=dir)
-        if expected_out is not None:
-            if is_regex:
-                if not re.match(expected_out, out):
-                    failed += 1
-                    print("FAILED running %s: wrong output" % t)
-                    print("Expected expression:")
-                    print(expected_out)
-                    print("Output:")
-                    print(out)
-                    print()
-                    continue
-            elif out != expected_out:
-                failed += 1
-                print("FAILED running %s: wrong output" % t)
-                for line in difflib.unified_diff(out.splitlines(True), expected_out.splitlines(True)):
-                    sys.stdout.write(line)
-                print()
-                continue
-        if expected_code != code:
-            failed += 1
-            print("FAILED running %s: wrong exit code, got %d but expected %d" % (t, code, expected_code))
-            print("Output:")
-            print(out)
-            print()
+    for t in tests:
+        path = {
+            "dir": os.path.dirname(t),
+            "file": os.path.basename(t),
+        }
+
+        logging.debug("Testing file: %s" % t)
+
+        curr = FbpTest(t)
+        if curr.should_skip(skips):
+            logging.debug("SKIPPED running %s: %s" % (t, curr.skip_msg))
+            status["skipped"] += 1
             continue
-        if expected_out is None and len(out) > 0:
-            print("IGNORING unexpected output from", t)
-            print(out)
-        passed += 1
-
-    print("%d tests passed." % passed)
-    if skipped > 0:
-        print("%d tests skipped." % skipped)
-    if failed > 0:
-        print("%d tests failed." % failed)
+
+        if not args.compiled:
+            cmd = cmd_prefix + [runner, path["file"]]
+            test_cwd = path["dir"]
+        else:
+            exe = os.path.join(args.compiled_dir, path["file"].replace(".fbp", ""))
+            cmd = cmd_prefix + [exe]
+            test_cwd = "./"
+
+        out, code = sh(cmd, cwd=test_cwd)
+        result = curr.show_result_status(out, code)
+        status[result] += 1
+        if result == "failed":
+            failures = True
+
+    for k,v in status.items():
+        if v > 0:
+            logging.warning("%d tests %s." % (v, k))
+
+    if failures:
         sys.exit(1)
+
+def get_skip_list(args):
+    result = []
+    for curr in args.skip:
+        if not curr.upper().startswith('SKIP_'):
+            logging.info("Invalid skip spec: %s" % curr)
+            continue
+
+        skip = getattr(FbpTest, curr.upper(), None)
+        if not skip:
+            logging.info("Unknown skip spec: %s" % curr)
+
+        result.append(skip)
+    return set(result)
+
+def query_tests(args):
```

Comment:
Maybe a generator expression could simplify this function (nitpick and bikeshedding in a single comment):

    skips = get_skip_list(args)
    return (FbpTest(t) for t in collect_tests(args.tests) if not should_skip(t))

And then just print the result outside:

   print(" ".join(query_tests(args))